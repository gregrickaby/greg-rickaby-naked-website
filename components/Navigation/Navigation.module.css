.nav {
  @apply relative;
}

.navContainer {
  @apply flex items-center justify-between;
}

.navButton {
  @apply flex lg:hidden;
}

.linkContainer {
  @apply absolute right-0 top-12 z-10 text-right;
  @apply flex-col gap-4 rounded-lg pb-8 pl-8 pr-0 pt-8;
  @apply lg:static lg:flex lg:flex-row lg:rounded-none lg:bg-transparent lg:shadow-none;
  @apply bg-white dark:bg-gray-950;
}

.link {
  @apply text-xl hover:underline;
}
