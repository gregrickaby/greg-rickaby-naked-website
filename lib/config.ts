import {
  IconBook,
  IconBrandAmazon,
  IconBrandBluesky,
  IconBrandCodepen,
  IconBrandFacebook,
  IconBrandFeedly,
  IconBrandGithub,
  IconBrandInstagram,
  IconBrandLinkedin,
  IconBrandThreads,
  IconBrandWordpress,
  IconBrandYoutube,
  IconRss
} from '@tabler/icons-react'

const config = {
  siteName: 'Greg Rickaby',
  siteDescription: 'Cameras and Code',
  siteUrl: 'https://gregrickaby.com',
  siteLogo: '/logo.webp',
  siteOgImage: '/og.webp',
  jobTitle: 'Technical Lead',
  email: 'greg@gregrickaby.com',
  intro: 'Disney Dad. Full-Stack Developer. Photographer. Author.',
  gravatarApiUrl:
    'https://en.gravatar.com/28af3e39c0a1fe4c31367c7e9a8bcac3.json',
  navigation: [
    {
      name: 'About',
      url: '/about'
    },
    {
      name: 'Blog',
      url: '/blog'
    },
    {
      name: 'Contact',
      url: '/contact'
    },
    {
      name: 'Media',
      url: '/media'
    },
    {
      name: 'Resources',
      url: '/resources'
    },
    {
      name: 'Resume',
      url: '/resume'
    }
  ],
  socials: [
    {
      name: 'Bluesky',
      url: 'https://bsky.app/profile/gregrickaby.bsky.social',
      icon: IconBrandBluesky
    },
    {
      name: 'Instagram',
      url: 'https://www.instagram.com/gregoryrickaby/',
      icon: IconBrandInstagram
    },
    {
      name: 'Threads',
      url: 'https://www.threads.net/@gregoryrickaby',
      icon: IconBrandThreads
    },
    {
      name: 'Facebook',
      url: 'https://www.facebook.com/gregrickaby',
      icon: IconBrandFacebook
    },
    {
      name: 'GitHub',
      url: 'https://github.com/gregrickaby/',
      icon: IconBrandGithub
    },
    {
      name: 'LinkedIn',
      url: 'https://linkedin.com/in/gregrickaby/',
      icon: IconBrandLinkedin
    },
    {
      name: 'WordPress',
      url: 'https://profiles.wordpress.org/gregrickaby/',
      icon: IconBrandWordpress
    },
    {
      name: 'Amazon Author',
      url: 'https://www.amazon.com/author/gregrickaby',
      icon: IconBrandAmazon
    },
    {
      name: 'Goodreads',
      url: 'https://www.goodreads.com/author/show/16999736.Greg_Rickaby',
      icon: IconBook
    },
    {
      name: 'CodePen',
      url: 'https://codepen.io/gregrickaby',
      icon: IconBrandCodepen
    },
    {
      name: 'YouTube',
      url: 'https://www.youtube.com/@GregRickaby',
      icon: IconBrandYoutube
    },
    {
      name: 'Feedly',
      url: 'https://feedly.com/i/subscription/feed%2Fhttps%3A%2F%2Fgregrickaby.com%2Ffeed.xml',
      icon: IconBrandFeedly
    },
    {
      name: 'RSS Feed',
      url: 'https://gregrickaby.com/feed.xml',
      icon: IconRss
    }
  ]
}

export default config
