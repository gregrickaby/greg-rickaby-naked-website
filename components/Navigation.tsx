'use client'

import config from '@/lib/config'
import {IconMenu, IconX} from '@tabler/icons-react'
import clsx from 'clsx'
import Link from 'next/link'
import {MouseEvent, useEffect, useRef, useState} from 'react'

/**
 * Navigation component.
 */
export default function Navigation() {
  const [isOpen, setIsOpen] = useState(false)
  const navRef = useRef<HTMLDivElement | null>(null)

  /**
   * Handle link click.
   */
  const handleLinkClick = () => {
    setIsOpen(false)
  }

  /**
   * Handle click outside.
   *
   * @param event - Mouse event.
   * @returns void
   */
  const handleClickOutside = (event: MouseEvent | Event) => {
    if (navRef.current && !navRef.current.contains(event.target as Node)) {
      setIsOpen(false)
    }
  }

  /**
   * Add event listener.
   */
  useEffect(() => {
    document.addEventListener('mousedown', handleClickOutside)
    return () => {
      document.removeEventListener('mousedown', handleClickOutside)
    }
  }, [])

  return (
    <nav ref={navRef} className="relative" data-testid="nav">
      <div className="flex items-center justify-between">
        <button
          aria-label="Toggle navigation"
          className="button flex lg:hidden"
          onClick={() => setIsOpen(!isOpen)}
        >
          {isOpen ? <IconX size={24} /> : <IconMenu size={24} />}
        </button>
      </div>
      <div
        className={clsx(
          'absolute right-0 top-12 z-10 flex-col gap-4 rounded-lg bg-white pb-8 pl-8 pr-0 pt-8 text-right lg:static lg:flex lg:flex-row lg:rounded-none lg:bg-transparent lg:shadow-none dark:bg-gray-950',
          {flex: isOpen, hidden: !isOpen}
        )}
        data-testid="nav-links"
      >
        {config.navigation.map((item) => (
          <Link
            className="text-xl hover:underline"
            href={item.url}
            key={item.name}
            onClick={handleLinkClick}
          >
            {item.name}
          </Link>
        ))}
      </div>
    </nav>
  )
}
