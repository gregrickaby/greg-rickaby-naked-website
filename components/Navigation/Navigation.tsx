'use client'

import config from '@/lib/config'
import {IconMenu, IconX} from '@tabler/icons-react'
import clsx from 'clsx'
import Link from 'next/link'
import {MouseEvent, useEffect, useRef, useState} from 'react'
import styles from './Navigation.module.css'

/**
 * Navigation component.
 */
export function Navigation() {
  const [isOpen, setIsOpen] = useState(false)
  const navRef = useRef<HTMLDivElement | null>(null)

  /**
   * Handle link click.
   */
  const handleLinkClick = () => {
    setIsOpen(false)
  }

  /**
   * Handle click outside.
   *
   * @param event - Mouse event.
   * @returns void
   */
  const handleClickOutside = (event: MouseEvent | Event) => {
    if (navRef.current && !navRef.current.contains(event.target as Node)) {
      setIsOpen(false)
    }
  }

  /**
   * Add event listener.
   */
  useEffect(() => {
    document.addEventListener('mousedown', handleClickOutside)
    return () => {
      document.removeEventListener('mousedown', handleClickOutside)
    }
  }, [])

  return (
    <nav ref={navRef} className={styles.nav} data-testid="nav">
      <div className={styles.navContainer}>
        <button
          aria-label="Toggle navigation"
          className={clsx('button', styles.navButton)}
          onClick={() => setIsOpen(!isOpen)}
        >
          {isOpen ? <IconX size={24} /> : <IconMenu size={24} />}
        </button>
      </div>
      <div
        className={clsx(styles.linkContainer, {flex: isOpen, hidden: !isOpen})}
        data-testid="nav-links"
      >
        {config.navigation.map((item) => (
          <Link
            className={styles.link}
            href={item.url}
            key={item.name}
            onClick={handleLinkClick}
          >
            {item.name}
          </Link>
        ))}
      </div>
    </nav>
  )
}
