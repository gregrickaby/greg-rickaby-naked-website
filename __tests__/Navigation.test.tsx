import Navigation from '@/components/Navigation'
import config from '@/lib/config'
import '@testing-library/jest-dom'
import {render} from '@testing-library/react'
import {axe} from 'jest-axe'
import {describe, expect, it} from 'vitest'

describe('Navigation', () => {
  it('should render', () => {
    const {container} = render(<Navigation />)
    expect(container).toBeInTheDocument()
  })

  it('it should render every link in the config', () => {
    const {getByTestId} = render(<Navigation />)
    const nav = getByTestId('nav')
    expect(nav.children.length).toBe(config.navigation.length)
  })

  it('should not have any accessibility issues', async () => {
    const {container} = render(<Navigation />)
    const results = await axe(container)
    expect(results).toHaveNoViolations()
  })

  it('should match snapshot', () => {
    const {getByTestId} = render(<Navigation />)
    const footer = getByTestId('nav')
    expect(footer).toMatchSnapshot()
  })
})
