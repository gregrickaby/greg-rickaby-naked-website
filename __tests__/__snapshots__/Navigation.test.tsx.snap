// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Navigation > should match snapshot 1`] = `
<nav
  class="relative"
  data-testid="nav"
>
  <div
    class="flex items-center justify-between"
  >
    <button
      aria-label="Toggle navigation"
      class="button flex lg:hidden"
    >
      <svg
        class="tabler-icon tabler-icon-menu "
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M4 8l16 0"
        />
        <path
          d="M4 16l16 0"
        />
      </svg>
    </button>
  </div>
  <div
    class="absolute right-0 top-12 z-10 flex-col gap-4 rounded-lg bg-white pb-8 pl-8 pr-0 pt-8 text-right lg:static lg:flex lg:flex-row lg:rounded-none lg:bg-transparent lg:shadow-none dark:bg-gray-950 hidden"
    data-testid="nav-links"
  >
    <a
      class="text-xl hover:underline"
      href="/"
    >
      Home
    </a>
    <a
      class="text-xl hover:underline"
      href="/about"
    >
      About
    </a>
    <a
      class="text-xl hover:underline"
      href="/blog"
    >
      Blog
    </a>
    <a
      class="text-xl hover:underline"
      href="/contact"
    >
      Contact
    </a>
    <a
      class="text-xl hover:underline"
      href="/cv"
    >
      CV
    </a>
    <a
      class="text-xl hover:underline"
      href="/media"
    >
      Media
    </a>
    <a
      class="text-xl hover:underline"
      href="/resources"
    >
      Resources
    </a>
  </div>
</nav>
`;
